compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs

```diff
@@ -524,6 +554,86 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         }
     }
 
+    /// Determines whether deref_to == <deref_from as Deref>::Target, and if so,
+    /// returns a prefix that should be added to deref_from as a suggestion.
+    fn should_deref_suggestion_on_mismatch(
+        &self,
+        cause: &ObligationCause<'tcx>,
+        param_env: ParamEnv<'tcx>,
+        deref_to: Ty<'tcx>,
+        deref_from: Ty<'tcx>,
+    ) -> Option<String> {
+        let tcx = self.infcx.tcx;
+
+        let deref_trait = tcx.lang_items().deref_trait()?;
+        let deref_mut_trait = tcx.lang_items().deref_mut_trait()?;
+        let deref_target = tcx.lang_items().deref_target()?;
+
+        // Count the number of references in deref_from, since we need
+        // that plus 1 to invoke Deref.
+        // This is also needed to ensure we aren't checking for Deref on a reference.
+        let mut deref_from_cur = deref_from;
+        // Extra required dereference to invoke Deref.
+        let mut num_refs = 1;
+        while let ty::Ref(_, inner_ty, _) = deref_from_cur.kind() {
+            deref_from_cur = *inner_ty;
+            num_refs += 1;
+        }
+
+        // <deref_from_cur as Deref>
+        let trait_ref = ty::TraitRef::new(tcx, deref_trait, [deref_from_cur]);
+        let obligation =
+            Obligation::new(tcx, cause.clone(), param_env, ty::Binder::dummy(trait_ref));
+        if !self.infcx.predicate_may_hold(&obligation) {
+            return None;
+        }
+
+        let ocx = ObligationCtxt::new(self.infcx);
+
+        // <deref_from_cur as Deref>::Target
+        let target = Ty::new_projection(tcx, deref_target, [deref_from_cur]);
```

Comment:
You may find it possible to simplify the logic if you use the `autoderef_steps` helper:
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/error_reporting/struct.TypeErrCtxt.html#structfield.autoderef_steps

Reply:
You shouldn't need to normalize anything if you use that helper. Just do something like `.find(|ty| can_eq(ty, target))`.